lang/rust/rustc-triple.mk

```diff
@@ -0,0 +1,94 @@
+# To enable rustc/cargo support in your packages, add:
+# include $(TOPDIR)/feeds/packages/lang/rustup/rustc-triple.mk
+# below include $(TOPDIR)/rules.mk
+#
+# You will need to add --host=$(RUSTC_TARGET_ARCH) to your packages
+# CONFIGURE_VARS to cross-compile for the target architecture
+
+RUSTC_ARCH_TARGETS:= \
+aarch64-unknown-linux-gnu aarch64-unknown-linux-musl \
+\
+arm-unknown-linux-gnueabi arm-unknown-linux-gnueabihf \
+arm-unknown-linux-musleabi arm-unknown-linux-musleabihf \
+\
+armv4t-unknown-linux-gnueabi \
+\
+armv5te-unknown-linux-gnueabi armv5te-unknown-linux-musleabi \
+\
+armv7-unknown-linux-gnueabi armv7-unknown-linux-gnueabihf \
+armv7-unknown-linux-musleabi \
+\
+hexagon-unknown-linux-musl \
+\
+i586-unknown-linux-gnu i586-unknown-linux-musl \
+\
+i686-unknown-linux-gnu i686-unknown-linux-musl \
+\
+mips-unknown-linux-gnu mips-unknown-linux-musl mips-unknown-linux-uclibc \
+\
+mips64-unknown-linux-gnuabi64 mips64-unknown-linux-muslabi64 \
+mips64-unknown-linux-gnuabi64sf mips64-unknown-linux-muslabi64sf \
+\
+mips64el-unknown-linux-gnuabi64 mips64el-unknown-linux-muslabi64 \
+\
+mipsel-unknown-linux-gnu mipsel-unknown-linux-musl \
+mipsel-unknown-linux-uclibc \
+\
+mipsisa32r6-unknown-linux-gnu mipsisa32r6el-unknown-linux-gnu \
+\
+mipsisa64r6-unknown-linux-gnuabi64 mipsisa64r6el-unknown-linux-gnuabi64 \
+\
+powerpc-unknown-linux-gnu powerpc-unknown-linux-musl \
+\
+powerpc64-unknown-linux-gnu powerpc64-unknown-linux-musl \
+\
+powerpc64le-unknown-linux-gnu powerpc64le-unknown-linux-musl \
+\
+riscv64gc-unknown-linux-gnu \
+\
+s390x-unknown-linux-gnu \
+\
+sparc-unknown-linux-gnu \
+\
+sparc64-unknown-linux-gnu \
+\
+thumbv7neon-unknown-linux-gnueabihf thumbv7neon-unknown-linux-musleabihf \
+\
+x86_64-unknown-linux-gnu x86_64-unknown-linux-musl
+
+CONFIG_HOST_SUFFIX:=$(shell cut -d"-" -f4 <<<"$(GNU_HOST_NAME)")
+RUSTC_HOST_ARCH:= \
+	$(strip $(foreach \
+		v, \
+		$(filter $(HOST_ARCH)-%, $(RUSTC_ARCH_TARGETS)), \
+		$(if $(findstring -$(CONFIG_HOST_SUFFIX:"%"=%),$v),$v) \
+		) \
+	)
+
+RUSTC_TARGET_ARCH_BASE:= \
+	$(strip $(foreach \
+		v, \
+		$(filter $(ARCH)-%, $(RUSTC_ARCH_TARGETS)), \
+		$(if $(findstring -$(CONFIG_TARGET_SUFFIX:"%"=%),$v),$v) \
+		) \
+	)
+
+# Check to see if it's a soft-float target
+ifeq ($(CONFIG_SOFT_FLOAT),y)
+RUSTC_TARGET_ARCH:=$(strip $(filter %sf, $(RUSTC_TARGET_ARCH_BASE)))
+else
+RUSTC_TARGET_ARCH:=$(strip $(filter-out %sf, $(RUSTC_TARGET_ARCH_BASE)))
+endif
```

Comment:
This whole situation for finding the host and target triple probably can be cleaner.  I know SOME targets are soft-float by default.. Some are Hard-float by default.  They aren't always named so you can tell (mipsel is soft-float by default, for example, but mips64 can be both, depending on the specific model).  I'm not sure which other triples are going to have this issue or others similar.  Thoughts and ideas welcome

Reply:
`CONFIG_SOFT_FLOAT` can be checked here.  See rules.mk for details.

Reply:
~~`CONFIG_SOFT_FLOAT` can be checked here. See rules.mk for details.~~

Does it the rust configure command support separate option for specifying float implementation flavor?

Reply:
> ~`CONFIG_SOFT_FLOAT` can be checked here. See rules.mk for details.~
> 
> Does it the rust configure command support separate option for specifying float implementation flavor?

No.. Soft-Float is set in the Triple defines 

This is why I went with rust-from-source rather than Rustup - Rustup defaults to statically linked and hard-float.

The issues that I'm going to face has to do with the way Rust standardizes their triples..

> grommish@norwits:~/openwrt/build_dir/hostpkg/rust/compiler/rustc_target/src/spec$ grep -r "+soft-float" *
> armv4t_unknown_linux_gnueabi.rs:            features: "+soft-float,+strict-align".to_string(),
> armv5te_unknown_linux_gnueabi.rs:            features: "+soft-float,+strict-align".to_string(),
> armv5te_unknown_linux_musleabi.rs:            features: "+soft-float,+strict-align".to_string(),
> armv7a_none_eabi.rs:        features: "+v7,+thumb2,+soft-float,-neon,+strict-align".to_string(),
> armv7_unknown_linux_gnueabi.rs:            features: "+v7,+thumb2,+soft-float,-neon".to_string(),
> armv7_unknown_linux_musleabi.rs:            features: "+v7,+thumb2,+soft-float,-neon".to_string(),
> i686_unknown_uefi.rs:    base.features = "-mmx,-sse,+soft-float".to_string();
> mips64_unknown_linux_muslabi64sf.rs:    base.features = "+mips64r2,+soft-float".to_string();
> mipsel_unknown_linux_musl.rs:    base.features = "+mips32r2,+soft-float".to_string();
> mipsel_unknown_linux_uclibc.rs:            features: "+mips32r2,+soft-float".to_string(),
> mipsel_unknown_none.rs:            features: "+mips32r2,+soft-float,+noabicalls".to_string(),
> mips_unknown_linux_musl.rs:    base.features = "+mips32r2,+soft-float".to_string();
> mips_unknown_linux_uclibc.rs:            features: "+mips32r2,+soft-float".to_string(),
> thumbv4t_none_eabi.rs:            features: "+soft-float,+strict-align".to_string(),
> x86_64_linux_kernel.rs:        "-mmx,-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-3dnow,-3dnowa,-avx,-avx2,+soft-float"
> x86_64_unknown_hermit_kernel.rs:        "-mmx,-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-3dnow,-3dnowa,-avx,-avx2,+soft-float"
> x86_64_unknown_uefi.rs:    base.features = "-mmx,-sse,+soft-float".to_string();
> 


You can see they call some soft-float in regular files, but then tag some with hf..  The mips64_unknown_linux_muslabi64sf.rs is my custom define, because mips64_unknown_linux_muslabi64.rs is NOT soft_float and my octeon3 chip requires it.

Reply:
I decided to  just implement custom rustc triples for Openwrt in a <arch>-openwrt-linux-<libc> format.